File: framework/languages/java/samples/java/DeliveryDispatch/Server/Tracking/src/main/java/systems/zlink/samples/deliverydispatch/server/tracking/handlers/DeliveryStatusChangedHandler.java
package systems.zlink.samples.deliverydispatch.server.tracking.handlers;

import systems.zlink.framework.actors.ZLinkActorClient;
import systems.zlink.framework.actors.ZLinkActorDirectory;
import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.configuration.EvidenceStore;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup("tracking")
public final class DeliveryStatusChangedHandler
    implements ZLinkRequestHandler<Messages.DeliveryStatusChangedReq, Messages.DeliveryStatusChangedRes> {
    private static final String SAMPLE_CUSTOMER_ID = "customer-1";
    private final EvidenceStore evidenceStore;
    private final ZLinkActorClient actors;
    private final ZLinkActorDirectory actorRefs;

    public DeliveryStatusChangedHandler(
        EvidenceStore evidenceStore,
        ZLinkActorClient actors,
        ZLinkActorDirectory actorRefs) {
        this.evidenceStore = evidenceStore;
        this.actors = actors;
        this.actorRefs = actorRefs;
    }

    @Override
    public CompletionStage<Messages.DeliveryStatusChangedRes> handle(
        Messages.DeliveryStatusChangedReq request,
        ZLinkMessageContext context) {
        evidenceStore.append(request);
        return actorRefs.find(SAMPLE_CUSTOMER_ID).thenCompose(found -> {
            var actor = found.orElseThrow(() -> new IllegalStateException(
                "customer actor not found: " + SAMPLE_CUSTOMER_ID));
            return actors.sendToActor(actor.actorId(), new Messages.DeliveryStatusUpdatedMsg(
                    request.deliveryId(), SAMPLE_CUSTOMER_ID, request.status(),
                    request.courierId(), request.occurredAt()))
                .submit()
                .thenApply(ignored ->
                    new Messages.DeliveryStatusChangedRes(
                        request.deliveryId(), request.status()));
        });
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/Tracking/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/tracking/handlers/DeliveryStatusChangedHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.tracking.handlers

import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.actors.ZLinkActorClient
import systems.zlink.framework.actors.ZLinkActorDirectory
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.DeliveryEvidenceStore
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusChangedRes
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusChangedReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusUpdatedMsg

@ZLinkHandlerGroup("tracking")
class DeliveryStatusChangedHandler(
    private val evidenceStore: DeliveryEvidenceStore,
    private val actors: ZLinkActorClient,
    private val actorRefs: ZLinkActorDirectory,
) : ZLinkSuspendingRequestHandler<DeliveryStatusChangedReq, DeliveryStatusChangedRes> {
    override suspend fun handle(
        request: DeliveryStatusChangedReq,
        context: ZLinkMessageContext,
    ): DeliveryStatusChangedRes {
        evidenceStore.append(request)
        val actorRef = actorRefs.find(request.customerId).await()
            .orElseThrow {
                IllegalStateException("customer actor not found: ${request.customerId}")
            }
        actors.sendToActor(
            actorRef.actorId,
            DeliveryStatusUpdatedMsg(
                deliveryId = request.deliveryId,
                customerId = request.customerId,
                status = request.status,
                courierId = request.courierId,
                occurredAt = request.occurredAt,
            ),
        )
            .submit()
        return DeliveryStatusChangedRes(request.deliveryId, request.status)
    }
}
